warehouse/schema/schema.go

```diff
@@ -78,56 +77,89 @@ type schema struct {
+	slogger logger.Logger,
 	statsFactory stats.Stats,
 	fetchSchemaRepo fetchSchemaRepo,
-) Handler {
+	schemaRepo schemaRepo,
+	stagingFileRepo stagingFileRepo,
+) (Handler, error) {
 	ttlInMinutes := conf.GetDurationVar(720, time.Minute, "Warehouse.schemaTTLInMinutes")
-	schema := &schema{
+	sh := &schema{
 		warehouse:                        warehouse,
-		log:                              logger.Child("schema"),
+		log:                              slogger.Child("schema"),
 		ttlInMinutes:                     ttlInMinutes,
-		schemaRepo:                       repo.NewWHSchemas(db),
+		schemaRepo:                       schemaRepo,
 		stagingFilesSchemaPaginationSize: conf.GetInt("Warehouse.stagingFilesSchemaPaginationSize", 100),
-		stagingFileRepo:                  repo.NewStagingFiles(db),
+		stagingFileRepo:                  stagingFileRepo,
 		fetchSchemaRepo:                  fetchSchemaRepo,
 		enableIDResolution:               conf.GetBool("Warehouse.enableIDResolution", false),
 		now:                              timeutil.Now,
 	}
-	schema.stats.schemaSize = statsFactory.NewTaggedStat("warehouse_schema_size", stats.HistogramType, stats.Tags{
+	sh.stats.schemaSize = statsFactory.NewTaggedStat("warehouse_schema_size", stats.HistogramType, stats.Tags{
 		"module":        "warehouse",
 		"workspaceId":   warehouse.WorkspaceID,
 		"destType":      warehouse.Destination.DestinationDefinition.Name,
 		"sourceId":      warehouse.Source.ID,
 		"destinationId": warehouse.Destination.ID,
 	})
-	return schema
+	// cachedSchema can be computed in the constructor
+	// we need not worry about it getting expired in the middle of the job
+	// since we need the schema to be the same for the entireduration of the job
+	whSchema, err := sh.schemaRepo.GetForNamespace(
+		ctx,
+		sh.warehouse.Source.ID,
+		sh.warehouse.Destination.ID,
+		sh.warehouse.Namespace,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("getting schema for namespace: %w", err)
+	}
+	if whSchema.Schema == nil {
+		sh.cachedSchema = model.Schema{}
+		return sh, nil
+	}
+	if whSchema.ExpiresAt.After(sh.now()) {
+		sh.cachedSchema = whSchema.Schema
+		return sh, nil
+	}
+	sh.log.Infon("Schema expired", obskit.DestinationID(sh.warehouse.Destination.ID), obskit.Namespace(sh.warehouse.Namespace), logger.NewTimeField("expiresAt", whSchema.ExpiresAt))
+	return sh, sh.fetchSchemaFromWarehouse(ctx)
 }
 
-func (sh *schema) IsSchemaEmpty(ctx context.Context) bool {
-	schema, err := sh.getSchema(ctx)
+func (sh *schema) fetchSchemaFromWarehouse(ctx context.Context) error {
+	start := sh.now()
+	warehouseSchema, err := sh.fetchSchemaRepo.FetchSchema(ctx)
+	if err != nil {
+		return fmt.Errorf("fetching schema: %w", err)
+	}
+	duration := math.Round((sh.now().Sub(start).Minutes() * 1000)) / 1000
+	sh.log.Infon("Fetched schema from warehouse", obskit.DestinationID(sh.warehouse.Destination.ID), obskit.Namespace(sh.warehouse.Type), logger.NewFloatField("timeTakenInMinutes", duration))
```

Comment:
minor: Why minutes and not seconds? Would values of few seconds get rounded as 0 if using minutes?

Reply:
> Why minutes and not seconds?

I believe we would be primarily interested in larger values, so I thought it would be better to use minutes instead of seconds.

> Would values of few seconds get rounded as 0 if using minutes?

No. `math.Round((sh.now().Sub(start).Minutes() * 1000)) / 1000` ensures value upto 3 decimals

Reply:
I would say can keep it seconds. As in using minutes do not give any advantage as such.